sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala

```diff
@@ -0,0 +1,119 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.benchmark
+
+import scala.util.Random
+
+import org.apache.spark.benchmark.Benchmark
+import org.apache.spark.sql.{functions, Column}
+import org.apache.spark.sql.catalyst.util.CollationFactory
+import org.apache.spark.sql.functions._
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * Benchmark to measure performance for joins. To run this benchmark:
+ * {{{
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class>
+ *        --jars <spark core test jar>,<spark catalyst test jar> <spark sql test jar>
+ *   2. build/sbt "sql/Test/runMain org.apache.spark.sql.execution.benchmark.CollationBenchmark"
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/Test/runMain <this class>"
+ *      Results will be written to "benchmarks/JoinBenchmark-results.txt".
+ * }}}
+ */
+
+object CollationBenchmark extends SqlBasedBenchmark {
+  private val collationTypes = Seq("UTF8_BINARY", "UTF8_BINARY_LCASE", "UNICODE", "UNICODE_CI")
+
+  def generateUTF8Strings(n: Int): Seq[UTF8String] = {
+    // Generate n UTF8Strings
+    Seq("ABC", "aBC", "abc", "DEF", "def", "GHI", "ghi", "JKL", "jkl",
+      "MNO", "mno", "PQR", "pqr", "STU", "stu", "VWX", "vwx", "YZ").map(UTF8String.fromString) ++
+    (18 to n).map(i => UTF8String.fromString(Random.nextString(i % 25))).sortBy(_.hashCode())
+  }
+
+  def benchmarkUTFString(collationTypes: Seq[String], utf8Strings: Seq[UTF8String]): Unit = {
+    val benchmark = collationTypes.foldLeft(
+      new Benchmark(s"collation unit benchmarks", utf8Strings.size, output = output)) {
+      (b, collationType) =>
+        b.addCase(s"equalsFunction - $collationType") { _ =>
+          val collation = CollationFactory.fetchCollation(collationType)
+          utf8Strings.slice(0, 20).foreach(s1 =>
+            utf8Strings.filter(s =>
+            collation.equalsFunction(s, s1).booleanValue()
+            )
+          )
+        }
+        b.addCase(s"collator.compare - $collationType") { _ =>
+          val collation = CollationFactory.fetchCollation(collationType)
+          utf8Strings.slice(0, 20).foreach(s1 =>
+            utf8Strings.sortBy(s =>
+              collation.comparator.compare(s, s1)
+            )
+          )
+        }
+        b.addCase(s"hashFunction - $collationType") { _ =>
+          val collation = CollationFactory.fetchCollation(collationType)
+          utf8Strings.map(s => collation.hashFunction.applyAsLong(s))
+        }
+        b
+    }
+    benchmark.run()
+  }
+
+  def collationBenchmarkFilterEqual(
```

Comment:
IMO, just having `benchmarkUTFString` would be great starting point. We can work on E2E benchmarks later.

Reply:
Do you mind if we keep them? I would tidy the code up if so. 

You can make the call about whether to omit them in this initial PR, but before you decide, you should first check out the result file -- the E2E tests are already pretty useful and revealing.

Reply:
For example. check out the data frame filtering speedup in performance for `UTF8_BINARY_LCASE` when going from Java 17 to 21 !

Reply:
@dbatomic let me know if removing the e2e changes is blocking approval, or just a nice-to-have?  I think this PR is all cleaned up and ready for another Review. LMK.

Reply:
It's not blocking the approval from my side.

I left a couple of additional comments. After that it should be green from my side.

Reply:
👍